JSON Schema for SPARQL results JSON

domel opened the pull request:
See also https://github.com/w3c/sparql-results-json/issues/9

domel committed:
JSON Schema for SPARQL results JSON

afs commented:
Has this been run on the examples in the document? Which tool did you use?

domel commented:
> Has this been run on the examples in the document? Which tool did you use?

Yes, I've tested all examples from spec. I use different online tools. All tests were passed.

afs commented on spec/result.json:
```diff
@@ -0,0 +1,73 @@
+                                },
+                                "xml:lang": {
+                                    "type": "string",
+                                    "maxLength": 35
```
Where does the 35 come from?

At the syntax level , there isn't a limit on the length of a lang tag (Turtle, SPARQL) except it must be at least one character). Even RDF/XML it's at least one character because xml:lang="" has a special meaning.

Tpt replied:
Idea: use the same regular expression as Turtle.

afs replied:
Could do, although the results capturing general information from the RDF abstract data model, which refers to BCP47.
The XML results schema does not have such restrictions.

The full regular expression is ghastly!
https://stackoverflow.com/questions/7035825/regular-expression-for-a-language-tag-as-defined-by-bcp47

The JSON schema regex definition is [specific subset of Javascript](https://json-schema.org/understanding-json-schema/reference/regular_expressions.html) and not language independent.

The value of having a schema is having a formal, executable description of the structure of the SPARQL JSON results described [section 4](https://w3c.github.io/sparql-results-json/spec/index.html#select-results-form) and [section 5](https://w3c.github.io/sparql-results-json/spec/index.html#ask-result-form).

Let's focus on structure which is recording what is already in the results document.

domel replied:
Yes, 35 chars are the maximum of a valid language tag in the context of  BCP47. It is possible to use regex, but it will be strongly complicated (and I think that won't be supported by JSON Schema validators). So I'd rather avoid regexs. Another option is to remove it.

afs commented on spec/result.json:
```diff
@@ -0,0 +1,73 @@
+                                    "type": "string",
+                                    "maxLength": 35
+                                },
+                                "datatype": {
```
Is it possible to make it so that `xml:lang` and `datatype` are exclusive?

Tpt replied:
I believe it's possible with `enum`. It should be also possible to state that `xml:lang` and `datatype` should only be used when `"type": "literal"` and that `"boolean"` and `"results"` are exclusive (and that `"header"/"vars"` must be used only when `"results"` is present)

domel replied:
@afs what about `rdf:langString`?

afs replied:
`rdf:langString` is implied if `xml:lang` is present.

`xsd:string` is implied if there is no `datatype` or `xml:lang` (which gives the bridge from RDF 1.0 to RDF 1.1).

afs commented on spec/result.json:
```diff
@@ -0,0 +1,73 @@
+                }
+            }
+        },
+        "boolean": {
```
"boolean" and "results" are exclusive.

domel replied:
yes. That's correct.

afs commented on spec/result.json:
```diff
@@ -0,0 +1,73 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
```
"draft".

The best we can do is to make this non-normative "for the assistance of developers".
(If there is a timescale for completion, I couldn't find it.)

afs commented on spec/result.json:
```diff
@@ -0,0 +1,73 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "type": "object",
```
By my reading of the JSON schema doc, this schema should have an `"$id"` .

domel committed:
exclude results and boolean, remove lang restriction

domel committed:
$id and exlusive dt and lang

domel commented:
The third commit addresses all comments.
- `$id` added, I assume that it is `http://www.w3.org/2023/sparql/results.json`
- exclude `results` and `boolean`
- exclude `datatype` and `xml:lang`
- remove restriction to `xml:lang` value

Tpt commented on spec/result.json:
```diff
@@ -0,0 +1,105 @@
+                                },
+                                "datatype": {
+                                    "type": "string",
+                                    "format": "uri"
```
I believe it should be `"iri"` and not `"uri"`. https://json-schema.org/understanding-json-schema/reference/string.html#uri-template

Tpt reviewed (commented):
Thank you for the updates.

I am a bit puzzed with the `"dependencies"` system because it does not easily allow to e.g. express that `"xml:lang"` should only be used on literals. I just made an attempt at using the `oneOf` construction instead to state that the result is either a boolean result or a table result and that terms can be either uri, blank node, language-tagged literals or other literal: https://gist.github.com/Tpt/cff5dabc331514ef6bf2d3f99c794837

Feel free to reuse the gist if you feel it's relevant.

domel commented:
> Thank you for the updates.
> 
> I am a bit puzzed with the `"dependencies"` system because it does not easily allow to e.g. express that `"xml:lang"` should only be used on literals. I just made an attempt at using the `oneOf` construction instead to state that the result is either a boolean result or a table result and that terms can be either uri, blank node, language-tagged literals or other literal: https://gist.github.com/Tpt/cff5dabc331514ef6bf2d3f99c794837
> 
> Feel free to reuse the gist if you feel it's relevant.

If I understand JSON Schema correct, it that case dependencies/not is the same as OneOf.

Tpt commented:
> If I understand JSON Schema correct, it that case dependencies/not is the same as OneOf.

Yes, the oneOf option covers what is currently expressed with dependencies/not (either results or boolean and either xml:lang or datatype). Hwoever it goes futhers with things like "no head/vars if boolean", "values must be type: literal to have xml:lang" and "if the value type is uri then the value is an iri". So, I believe it covers more constraints than the dependencies/not encoding.

domel commented:
> Yes, the oneOf option covers what is currently expressed with dependencies/not (either results or boolean and either xml:lang or datatype). Hwoever it goes futhers with things like "no head/vars if boolean", "values must be type: literal to have xml:lang" and "if the value type is uri then the value is an iri". So, I believe it covers more constraints than the dependencies/not encoding.

@Tpt OK, I tried to validate examples from the spec using your JSON Schema and e.g. [JSON Schema Validator](https://www.jsonschemavalidator.net/), and it reports errors.

domel committed:
change uri to iri

Tpt commented:
> @Tpt OK, I tried to validate examples from the spec using your JSON Schema and e.g. [JSON Schema Validator](https://www.jsonschemavalidator.net/), and it reports errors.

Yes, I made an error with literals. I just fixed it. Sorry for that.

domel committed:
dependences/not to anyOf

@Tpt proposal

domel commented:
@Tpt thanks. I've committed changes. Please check.

afs commented:
Grammar looks and something people can test.

The other piece will be a change to the HTML.

Like [sparql-results-xml, section "XML Schemas](https://w3c.github.io/sparql-results-xml/spec/index.html#schemas) a short section that points to the file.

I don't think inlining, even in an appendix, is particularly useful. A separate file has it's own URL is better.

rubensworks commented on spec/result.json:
```diff
@@ -0,0 +1,156 @@
+                            }
+                        }
+                    },
+                    "additionalProperties": false
```
Do we really want to disallow additional properties here and later?
This explicitly forbids extension points, which is not forbidden in the current spec AFAICS.

Tpt replied:
Great point!

Proposal:
- root is extensible
- `head` is extensible
- `results` is maybe extensible (the spec states <q>The value of the "results" member is an object with a single key, "bindings".</q> but I don't see the harm of making it extensible)
- RDF terms are not extensible

rubensworks replied:
> RDF terms are not extensible

I would even allow those to be extensible (since that is not explicitly disallowed in the spec at the moment).

My suggestion would be to remove all occurences of `additionalProperties` throughout the whole schema.

domel committed:
small fix

pchampin commented:
could this be labelled as spec:enhancement ?

afs commented:
From the telcon (2023-03-30), this PR can be merged.

That does not mean that the schema is considered fixed and final.
